cli/internal/packager/compose.go

```diff
@@ -0,0 +1,108 @@
+package packager
+
+import (
+	"github.com/defenseunicorns/zarf/cli/config"
+	"github.com/defenseunicorns/zarf/cli/internal/message"
+	"github.com/defenseunicorns/zarf/cli/internal/packager/validate"
+	"github.com/defenseunicorns/zarf/cli/internal/utils"
+	"github.com/defenseunicorns/zarf/cli/types"
+)
+
+func GetComposedComponents() (components []types.ZarfComponent) {
+	for _, component := range config.GetComponents() {
+		// Check for standard component.
+		if !hasComposedPackage(&component) {
+			// Append standard component to list.
+			components = append(components, component)
+		} else if shouldComposePackage(&component) { // Validate and confirm inclusion of imported package.
+			// Expand and add components from imported package.
+			importedComponents := getSubPackageAssets(component)
+			components = append(components, importedComponents...)
+		}
+	}
+	// Update the parent package config with the expanded sub components.
+	// This is important when the deploy package is created.
+	config.SetComponents(components)
+	return components
+}
+
+// Returns true if import field is populated.
+func hasComposedPackage(component *types.ZarfComponent) bool {
+	return component.Import != types.ZarfImport{}
+}
+
+// Validates and confirms inclusion of imported package.
+func shouldComposePackage(component *types.ZarfComponent) bool {
+	return canComposePackage(component) && (config.DeployOptions.Confirm || component.Required || ConfirmOptionalComponent(*component))
+}
+
+// Validates the sub component, errors out if validation fails.
+func canComposePackage(component *types.ZarfComponent) bool {
```

Comment:
There's a bit of mixed approaches with how this function is handling outcomes. It returns `true` if nothing goes wrong but instead of returning a `false` it will only ever end the process with a fatal message. There's nothing logically wrong with this but I think it makes it a little harder to understand.

Maybe instead of returning an error, `ValidateImportPackage()` could just immediately go end the process with a fatal message like how `./cli/internal/utils/shasum.ValidateSha256Sum()` does?

Reply:
I would argue that it makes it less explicit and also violates open closed. It is a pattern I am seeing throughout the code that there are little to pure functions or classes and instead of returning values that are handled when called giving the ability to iterate or change behavior, we are hard coding and using direct manipulation and violating singleton patterns. By returning an error and making it explicit where it is used that we are failing out. it leaves the code open for change. By terminating within the function I am simple making void calls that make little to no sense when read in the code. The proposed change: 

```go
func canComposPackage(component *types.ZarfComponent) bool {
  	validate.ValidateImportPackage(component)
	return true
}
```
Is hiding the fatal logic. 

But I am open to changing it.

Reply:
I'm also not super strong in my opinion. But if ValidateImportPackage() were to just fatal then there would be no need for the function canComposePakcage() and the validation function would just be called earlier.

My point was more that having a function that claims to return a bool but really only returns true is just delaying which layer you're making the void call at.

Reply:
Actually I think I see your point now, there may be times where we want to check the validation without stopping the program. I guess the thing that caught my eye was the way we were only really ever returning 'true' which seems to make the return value meaningless to check.

Reply:
Yeah it's not my favorite but it avoids if/else hell and makes it composable by wrapping like that. Also it's likely there will be more ways to escape that method when we get to overriding imported components or even pulling from a remote source. 

I see your point with the called earlier, but when composed it makes it extremely useful and still gets called first. 
```
// Validates and confirms inclusion of imported package.
func shouldComposePackage(component *types.ZarfComponent) bool {
	return canComposePackage(component) && (config.DeployOptions.Confirm || component.Required || ConfirmOptionalComponent(*component))
}
```
By providing boolean logic I am actually allowing it to be called earlier otherwise I would have to check that the component is composable then inside the if I would have to run the function next. Because it is composed into a larger check, where the other options don't really fit within the scope of validation and are instead user preference based.

Reply:
And likely I should pull out those other 3 boolean checks into a function to clean that up a bit. Im going to do that now.